src/shogun/mathematics/graph/Graph.cpp

```diff
@@ -0,0 +1,185 @@
+#include <shogun/mathematics/graph/Graph.h>
+#include <shogun/mathematics/graph/Tensor.h>
+#include <shogun/mathematics/graph/ops/Input.h>
+
+#include <unordered_set>
+
+using namespace shogun;
+
+Graph::Graph(
+    const std::vector<std::shared_ptr<Input>>& inputs,
+    const std::vector<std::shared_ptr<Node>>& outputs)
+{
+	m_inputs = inputs;
+	m_outputs = outputs;
+}
+
+void Graph::build()
+{
+	auto unordered_nodes = check_fully_connected(m_inputs, m_outputs);
+	build_backend_graph(unordered_nodes);
+}
+
+std::map<std::shared_ptr<Node>, Graph::STATUS> Graph::check_fully_connected(
+    const std::vector<std::shared_ptr<Input>>& inputs,
+    const std::vector<std::shared_ptr<Node>>& outputs)
+{
+	std::deque<std::shared_ptr<Node>> nodes_to_check;
```

Comment:
couldn't we simply use the iterator ctor of std::deque?
```
std::deque<std::shared_ptr<Node>> nodes_to_check(outputs.begin(), outputs.end())
```

Reply:
ah, not used to  using deque, didn't realise there was this!